### PR TITLE
2019-05-23 Meeting minute

### LINKED ISSUE #20: Move to a single d2-ui package
It seems like we are leaning towards moving to a single d2-ui npm package with named exports for each of the components rather than components being maintained in separate packages.

If it is decided to go for this, how should it be done?

### LINKED ISSUE #23: Which future syntax do we want to use in our js?
I was going to open a jira issue for this, but it seemed slightly too philosophical for that. I'm wondering what level of the ecmascript proposals we officially want to use in our apps. As in, do we allow any and all future syntax? Do we only use syntax that's highly likely to not have to be changed?

I'm asking because we're using some future syntax in our apps, but I'm not aware of any dhis2 stance on what we do and don't use. So it seems to be up to the app devs. I think it would be good to have some clarity on that by sharing/discussing our stance.

For reference, the stages:

> Stage 0 “strawman” — The starting point for all proposals. These can change significantly before advancing to the next stage. There is no acceptance criteria and anyone can make a new proposal for this stage. There doesn’t need to be any implementation and the spec isn’t held to any standard. This stage is intended to start a discussion about the feature. There are currently over twenty stage 0 proposals.

> Stage 1 “proposal” — An actual formal proposal. These require a “champion”(i.e. a member of TC39 committee). At this stage the API should be well thought out and any potential implementation challenges should be outlined. At this stage, a polyfill is developed and demos produced. Major changes might happen after this stage, so use with caution. Proposals at this stage include the long-awaited Observables type and the Promise.try function.

> Stage 2 “draft” — At this stage the syntax is precisely described using the formal TC39 spec language. Minor editorial changes might still happen after this stage, but the specification should be complete enough not to need major revisions. If a proposal makes it this far, it’s a good bet that the committee expects the feature to be included eventually.

> Stage 3 “candidate” — The proposal has approved and further changes will only occur at the request of implementation authors. Here is where you can expect implementation to begin in JavaScript engines. Polyfills for proposals at this stage are safe to use without worry.

> Stage 4 “finished” — Indicates that the proposal has been accepted and the specification has been merged with the main JavaScript spec. No further changes are expected. JavaScript engines are expected to ship their implementations. As of October 2017 there are nine finished proposals, most notably async functions.

> From https://blog.logrocket.com/interesting-ecmascript-2017-proposals-163b787cf27c

Related to this we might also have to discuss APIs actually.

### LINKED ISSUE #24: Use github's stale bot / autoresponder / templates to help newcomers contribute
Moved from slack: regarding our discussion last week, about the jira vs github issues. I was thinking of resolving that in the following manner:

1. Add https://github.com/probot/stale bot to auto-close stale github issues and PRs after a period of no activity (we have a lot of those)
2. Add https://github.com/probot/autoresponder to let people know that (and how) they should post on Jira, as a friendly reminder
3. Add a .github repo for our org, so that we only have to configure our issue templates once (which autoresponder will use)

Number 3. also serves as a good default for other community health files (see https://help.github.com/en/articles/creating-a-default-community-health-file-for-your-organization), which can be overridden in each individual repo if necessary. It allows us to define issue templates, pr templates, etc. To make the project more friendly to newcomers and external contributors.

### LINKED ISSUE #28: How do we add cli-style to repos?
@varl and I were talking [here](https://github.com/dhis2/d2/pull/200#pullrequestreview-211760094) about how to implement cli-style for a repo. I tried starting from a common commit for the settings-app, adding cli-style and then merging to the five (master and 4 back) versions that we're actively using.

This leads to a bunch of conflicts, and I'm a bit worried that that might actually introduce more problems than it solves. Alternative is to just implement for master, which will make merging to master more difficult for our merge-base workflow.

So I'm wondering what others think we should do. Any perspectives on what the best approach would be? For reference, below two PRs that add cli-style to an app.

- https://github.com/dhis2/d2/pull/200
- https://github.com/dhis2/settings-app/pull/132

### LINKED ISSUE #44: Meeting agenda 2019-05-23
# Agenda

## #28 How do we add cli-style to repos?

This discussion is about cli-style, and how to introduce our code standards into a repository in the best way. The options discussed in #28 all have their drawbacks. The discussion about how to, and if, we should backport applies to any large refactoring to a code base, e.g.:

- Un-ejecting react-scripts.
- Porting to the AppShell.
- Introducing the code standards.

[Proposition from discussion](https://github.com/dhis2/notes/issues/28#issuecomment-484020882):

- We reformat all files directly.
- We do not backport style application and other major refactors such as un-ejecting
- We accept that backports become harder to automatically apply.
- We think that moving forward on master is more important than trying to drag 5 branches along for the ride.

## #24 Use github's stale bot / autoresponder / templates to help newcomers contribute

The stale-bot and some templates have been added to keep our issues clean and up to date. Autoresponder was not added on account of lack of configuration, and would lead bots to reply to bots.

The primary configuration is here: https://github.com/dhis2/.github
 
- Global configuration: https://github.com/dhis2/.github/blob/master/.github/stale.yml
- Repo configuration: https://github.com/dhis2/maintenance-app/blob/master/.github/stale.yml

Other files for [GitHub features](https://help.github.com/en/articles/setting-up-your-project-for-healthy-contributions):

- https://github.com/dhis2/.github/blob/master/CODE_OF_CONDUCT.md
     From: https://www.contributor-covenant.org/
- https://github.com/dhis2/.github/blob/master/CONTRIBUTING.md
- https://github.com/dhis2/.github/blob/master/SUPPORT.md
- We could potentially add PR and issue templates if we can figure out global templates for those.

## #20 Move to a single `d2-ui` package

Let's talk about the future of `d2-ui`. At its core, the problem is that the components are coupled to `d2`, which in turn is coupled to the version of dhis2, so there is complexity even when considering `d2-ui` in relation to its immediate dependencies and becomes harder to manage once we start to consider backports, and again when it's put into an App.

Either way we need to discuss `d2-ui` in relation to `app-runtime`, `ui-core`, and `ui-widgets` so outlining what they are and do for the record might be beneficial:

- `d2-ui` contains components coupled (there are some exceptions) with `d2`, and use `d2` models to get data from dhis2 to render.

- `ui-core` contains primitive components which can be understood as the basic LEGO blocks to build interfaces:

> (@hendrik) Anything containing *DHIS2 or domain specific business logic* should not go into `ui-core`

> (@amcgee) Another way to think about the distinction - nothing in `ui-core` should care that it’s dealing with DHIS2 data.  It’s a DHIS2-themed generic UI framework

- `ui-widgets` contains components which _are_ specific to DHIS2 and can deal with DHIS2 data and business logic. E.g. the `HeaderBar`. These are built using a data provider pattern and inversion of control in relation to how the components declare their data dependencies.

    HeaderBar example: https://github.com/dhis2/ui-widgets/blob/master/src/HeaderBar/index.js#L22-L35

    The App can then set up a "connected" `DataProvider` to handle those requests (or returned cached data, etc.), or pass through custom data, which is useful for e.g. [demos](https://github.com/dhis2/ui-widgets/blob/master/stories/HeaderBar.stories.js#L78-L80).

- `app-runtime` contains the runtime dependencies necessary for a DHIS2 application, for example the `DataProvider`.

So what does a single d2-ui package look like?

If done today something like:

- Delete `forms` (https://github.com/search?l=JSON&q=d2-ui-forms+org%3Adhis2&type=Code)
- Delete `core` and use a MUI / `ui-core` hybrid
- Delete `header-bar`
- Immediately reimplement `table` (https://github.com/search?l=JSON&q=d2-ui-table+org%3Adhis2&type=Code)
- Port `rich-text` to `ui-widgets` (https://github.com/dhis2/d2-ui/tree/master/packages/rich-text)

```
src
├── expression-manager
├── favorites-dialog
├── file-menu
├── formula-editor
├── group-editor
├── icon-picker
├── interpretations
├── legend
├── mentions-wrapper
├── org-unit-dialog
├── org-unit-select
├── org-unit-tree
├── period-picker
├── period-selector-dialog
├── sharing-dialog
├── translation-dialog
└── index.js
```

### PR DESCRIPTION
**Pull request by varl**
Closes #28 
Closes #24 
Closes #20 
Closes #44

### CODE REVIEW

**Review by ghost — APPROVED** [review]
Looks good! Nice meeting 👍

**Review by HendrikThePendric — APPROVED** [review]

**Review by amcgee — APPROVED** [review]
LGTM

File: decisions/2019/05/23-d2-ui-stale-styles-etc.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+- Think about if the app is suitable for feature toggling as we are
+  using it in a couple of places and getting some experience with good
+  results.
+
+- Be clear about pushing back on "backport everything, all the time",
+  especially as we move forward on continuous app deployments.
+
+**Conclusion**: We accept the proposal outlined.
+
+## #23 Use github's stale bot / autoresponder / etc.
```
Comment (erikarenhill) [inline decisions/2019/05/23-d2-ui-stale-styles-etc.md]: The correct issue for the stale bot / autoresponder is actually #24 and not #23

Comment (varl) [reply decisions/2019/05/23-d2-ui-stale-styles-etc.md]: nice, good catch.

**Review by erikarenhill — APPROVED** [review]

**Pull request merged**